src/cargo/core/compiler/layout.rs

```diff
@@ -146,13 +148,42 @@ impl Layout {
         if let Some(target) = target {
             root.push(target.short_name());
         }
+        // We need root to exist before we do the tempdir/rename dance inside it.
+        if !root.as_path_unlocked().exists() {
+            root.create_dir()?;
+        }
+        let dest_base = dest;
         let dest = root.join(dest);
         // If the root directory doesn't already exist go ahead and create it
         // here. Use this opportunity to exclude it from backups as well if the
         // system supports it since this is a freshly created folder.
+        //
+        // We do this in two steps (first create a temporary directory and exlucde
+        // it from backups, then rename it to the desired name. If we created the
+        // directory directly where it should be and then excluded it from backups
+        // we would risk a situation where cargo is interrupted right after the directory
+        // creation but before the exclusion the the directory would remain non-excluded from
+        // backups because we only perform exclusion right after we created the directory
+        // ourselves.
         if !dest.as_path_unlocked().exists() {
-            dest.create_dir()?;
-            exclude_from_backups(dest.as_path_unlocked());
+            // We need the tempdir created in root instead of $TMP, because only then we can be
+            // easily sure that rename() will succeed (the new name needs to be on the same mount
+            // point as the old one).
+            let tempdir = TempFileBuilder::new()
+                .prefix(dest_base)
+                .tempdir_in(root.as_path_unlocked())?;
+            exclude_from_backups(&tempdir.path());
+            // Previously std::fs::create_dir_all() (through paths::create_dir_all()) was used
+            // here to create the directory directly and fs::create_dir_all() explicitly treats
+            // the directory being created concurrently by another thread or process as success,
+            // hence the check below to follow the existing behavior. If we get an error at
+            // rename() and suddently the directory (which didn't exist a moment earlier) exists
+            // we can infer from it it's another cargo process doing work.
+            if let Err(e) = fs::rename(tempdir.path(), dest.as_path_unlocked()) {
+                if !dest.as_path_unlocked().exists() {
```

Comment:
Could this perhaps do what `create_dir_all` generally does by testing the error code returned?

Reply:
I wanted to do that but then found out there's no `ErrorKind` corresponding to the failure mode we expect in this case.

If `dest` exists and is empty `rename` will succeed, but when there are any files in that directory already (and we expect there to be, at least `CACHEDIR.TAG` with this patch) on Mac and Linux (not sure about other platforms) we get `ENOTEMPTY` and this gets reported as `ErrorKind::Other`. Extracting errno from the error seems easy (`Error::raw_os_error()`?) but I don't know what a platform-independent way of performing the verification of that error code would be.

After your comment I looked at how `create_dir_all()` does this in hope it'd help, but I think it does the same thing as my code (the last time of the listing):

```rust
// DirBuilder::create_dir_all()
    fn create_dir_all(&self, path: &Path) -> io::Result<()> {
        if path == Path::new("") {
            return Ok(());
        }

        match self.inner.mkdir(path) {
            Ok(()) => return Ok(()),
            Err(ref e) if e.kind() == io::ErrorKind::NotFound => {}
            Err(_) if path.is_dir() => return Ok(()),
(...)
```
(from https://github.com/rust-lang/rust/blob/1557fb031b272b4c5bfcc7de5df7eddc7b36a584/src/libstd/fs.rs#L2223)

Reply:
Ah ok, in that case this seems reasonable!